### ISSUE TITLE
Promises in where queries

### ISSUE THREAD

**Issue by bs1180** [join-monster/join-monster#41]
If I use a promise in a where query, it gets stringified:
```
where: async (courseTable, args, { viewer }) => {
  await permissionsCheck(viewer)
  return `${courseTable}.owner_id = viewer.id`
}
where: (courseTable, args, { viewer }) => {
   return Promise.resolve('')
}
```
Gets turned into:
`...WHERE "course".id = 1 AND [object Promise]\nORDER BY ...`

Thanks again for all your work on this.

**Comment by acarl005** [comment]
@bs1180 Thanks again for the initiative in submitting a PR!

I'm not yet convinced that the `where` function *should* be doing anything asynchronous. What is your use case?

Our thinking was that external information is to be retrieved prior to calling `joinMonster` and passed in via the context object, which is the third parameter to the where function. Is this not suitable for your use case?

**Comment by bs1180** [comment]
@acarl005 My use case is checking permissions and then modifying the query before it's run to ensure only the right information is returned. Here's an pseudo example of what it looked like before join-monster:

```
UserType:
  name: {
    type: GraphQLString
  },
  posts: {
    type: new GraphQLList(Post),
    resolve: async (root, args, { viewer }) => {
      let query = Knex('posts').where({ author_id: root.id })
      if (viewer.id === root.id) { // a user can see all his own posts
        query = query.andWhereIn({ status: ['public', 'draft'] })
      } else if (viewer.isAdmin()) { // an admin can see everything
        query = query.andWhereIn({ status: ['public', 'draft', 'deleted'] })
      } else if (await viewer.isFriend(root.id)) { // friends can see the last month of posts
        query = query.andWhere({ status: 'public', datePublished < 1 month })
      } else { // everyone else can only see the last week
        query = query.andWhere({ status: 'public', dataPublished < 1 week})
      }
      return query
  }
}

```

Using async methods like isFriend() lets me minimize unnecessary lookups, and I'm not sure it's feasible to fetch all possible data in advance so that it could be a synchronous method. So far I've been trying to just replicate this technique with join monster, but would happily try another approach if you have any suggestions.

**Comment by acarl005** [comment]
You're right. It can be impractical, intractable, or ungainly to request all the information in a resolver in order to do a permission check on a resource. The `where` function may be the most suitable place for this logic. It's not only a good encapsulation of logic, it prevents overfetching - one of the core goals of join monster.

These changes have been published in v0.6.0.

Thanks again for your contributions, @bs1180!
